edsnlp/tune.py

```diff
@@ -0,0 +1,400 @@
+            else:
+                raise ValueError(
+                    f"Unknown parameter type '{p_type}' for hyperparameter '{p_name}'."
+                )
+        else:
+            value = values[p_name]
+
+        current_config = config
+        for key in p_path[:-1]:
+            if key not in current_config:
+                raise KeyError(f"Path '{key}' not found in config.")
+            current_config = current_config[key]
+        current_config[p_path[-1]] = value
+
+    kwargs = Config.resolve(config["train"], registry=registry, root=config)
+    return kwargs, config
+
+
+def objective_with_param(config, tuned_parameters, trial):
+    kwargs, _ = update_config(config, tuned_parameters, trial=trial)
+    seed = random.randint(0, 2**32 - 1)
+    set_seed(seed)
+    try:
+        nlp = train(**kwargs, trial=trial)
+    except optuna.TrialPruned:
+        logger.info("Trial pruned")
+        raise
+    scorer = GenericScorer(**kwargs["scorer"])
+    val_data = kwargs["val_data"]
+    return scorer(nlp, val_data)["ner"]["micro"]["f"]
+
+
+def optimize(config_path, tuned_parameters, n_trials, study=None):
+    def objective(trial):
+        return objective_with_param(config_path, tuned_parameters, trial)
+
+    if not study:
+        study = optuna.create_study(
+            direction="maximize",
+            pruner=MedianPruner(n_startup_trials=5, n_warmup_steps=2),
+        )
+    study.optimize(objective, n_trials=n_trials)
+    return study
+
+
+def process_results(study, output_dir):
+    importances = compute_importances(study)
+    best_params = study.best_trial.params
+
+    logger.info(f"Best trial: {study.best_trial.number}")
+    logger.info(f"Value: {study.best_trial.value}")
+    logger.info(f"Params: {best_params}")
+    logger.info(f"Importances: {importances}")
+
+    os.makedirs(output_dir, exist_ok=True)
+    results_file = os.path.join(output_dir, "results_summary.txt")
+    with open(results_file, "w") as f:
+        f.write("Best Trial Summary\n")
+        f.write("==================\n")
+        f.write(f"Best trial: {study.best_trial.number}\n")
+        f.write(f"\nValue: {study.best_trial.value}\n")
+        f.write("\nParams:\n")
+        for key, value in best_params.items():
+            f.write(f"  {key}: {value}\n")
+        f.write("\nImportances:\n")
+        for key, value in importances.items():
+            f.write(f"  {key}: {value}\n")
+
+    os.makedirs(output_dir, exist_ok=True)
+    vis.plot_optimization_history(study).write_html(
+        os.path.join(output_dir, "optimization_history.html")
+    )
+    vis.plot_parallel_coordinate(study).write_html(
+        os.path.join(output_dir, "parallel_coordinate.html")
+    )
+    vis.plot_contour(study).write_html(os.path.join(output_dir, "contour.html"))
+    vis.plot_edf(study).write_html(os.path.join(output_dir, "edf.html"))
+    vis.plot_timeline(study).write_html(os.path.join(output_dir, "timeline.html"))
```

Comment:
Can we check if plotly is installed before launching the full tuning ? I foresee the case where a user has installed optuna manually and waits for several hours before the script crashes due to missing plotly. So either check before and fail, or, check before, warn if missing plotly, and wrap this in a try_except to avoid an ImportError exception